Add tutorials for migration from v1beta1 to v2. (#1982)
kyma-project/api-gateway: docs/user/tutorials/01-80-migrate-v1beta1-to-v2/01-84-migrate-oauth2-v1beta1-to-v2.md

## Prerequisites

* You have a deployed workload with the Istio and API Gateway modules enabled.
* To use the CLI instructions, you must have [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) and [curl](https://curl.se/) installed.
* You have a JSON Web Token. See [Obtain a JWT](../01-51-get-jwt.md).
* You have obtained the configuration of the APIRule in version `v1beta1` to be migrated. See [Retrieve the complete **spec**, including the **rules** field, of an APIRule in version `v1beta1`](./01-81-retrieve-v1beta1-spec.md).
* The workload exposed by the APIRule in version `v2` must be a part of the Istio service mesh.

## Steps

> [!NOTE] 
> In this example, the APIRule `v1beta1` was created with the **oauth2_introspection** handler, so the migration targets an APIRule `v2` using the **extAuth** handler. To illustrate the migration, the HTTPBin service is used, exposing the `/anything` and `/.*` endpoints. The HTTPBin service is deployed in its own namespace, with Istio enabled, ensuring the workload is part of the Istio service mesh.

> 1. Obtain a configuration of the APIRule in version `v1beta1` and save it for further modifications. For instructions, see [Retrieve the complete **spec**, including the **rules** field, of an APIRule in version `v1beta1`](./01-81-retrieve-v1beta1-spec.md). Below is a sample of the retrieved **spec** in YAML format for an APIRule in `v1beta1`:

Reviewer: ```suggestion
1. Obtain a configuration of the APIRule in version `v1beta1` and save it for further modifications. For instructions, see [Retrieve the Complete **spec** of an APIRule in Version `v1beta1`](./01-81-retrieve-v1beta1-spec.md). 

   See a sample of the retrieved **spec** in the YAML format:
```

```yaml
host: httpbin.local.kyma.dev
service:
  name: httpbin
  namespace: test
  port: 8000
gateway: kyma-system/kyma-gateway
rules:
  - path: /anything
    methods:
      - POST
    accessStrategies:
      - handler: oauth2_introspection
        config:
          introspection_request_headers:
            Authorization: Basic {ENCODED_CREDENTIALS}
          introspection_url: https://{IAS_TENANT}.accounts.ondemand.com/oauth2/introspect
          required_scope:
            - write
  - path: /.*
    methods:
      - GET
    accessStrategies:
      - handler: oauth2_introspection
        config:
          introspection_request_headers:
            Authorization: Basic {ENCODED_CREDENTIALS}
          introspection_url: https://{IAS_TENANT}.accounts.ondemand.com/oauth2/introspect
          required_scope:
            - read
```

> Above configuration uses the **oauth2_introspection** handler to expose HTTPBin `/anything` and `/.*` endpoints.

Reviewer: ```suggestion
The above configuration uses the **oauth2_introspection** handler to expose HTTPBin service's `/anything` and `/.*` endpoints.
```

> 2. In order for the `extAuth` handler in APIRule `v2` to work you must first deploy a service that will act as external authorizer for Istio. This tutorial uses [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) with an OAuth2.0 complaint authorization server supporting OIDC discovery with following configuration:

Reviewer: ```suggestion
2. In order for the `extAuth` handler in APIRule `v2` to work, you must first deploy a service that will act as an external authorizer for Istio. This tutorial uses [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) with an OAuth2.0 compliant authorization server supporting OIDC discovery.

   1. Replace the placeholders and create the `values.yaml` file with the oauth2-proxy configuration.
```
Maybe we could add substeps for step 2?

```yaml
cat <<EOF > values.yaml
config:
  clientID: {CLIENT_ID}
  clientSecret: {CLIENT_SECRET}
  cookieName: ""
  cookieSecret: {COOKIE_SECRET}

extraArgs:
  auth-logging: true
  cookie-domain: "{DOMAIN_TO_EXPOSE_WORKLOADS}"
  cookie-samesite: lax
  cookie-secure: false
  force-json-errors: true
  login-url: static://401
  oidc-issuer-url: {OIDC_ISSUER_URL}
  pass-access-token: true
  pass-authorization-header: true
  pass-host-header: true
  pass-user-headers: true
  provider: oidc
  request-logging: true
  reverse-proxy: true
  scope: "{TOKEN_SCOPES}"
  set-authorization-header: true
  set-xauthrequest: true
  skip-jwt-bearer-tokens: true
  skip-oidc-discovery: false
  skip-provider-button: true
  standard-logging: true
  upstream: static://200
  whitelist-domain: "*.{DOMAIN_TO_EXPOSE_WORKLOADS}:*"
EOF
```

> The example above shows the configuration of oauth2-proxy with the following parameters: `CLIENT_SECRET`, `CLIENT_ID`, and `OIDC_ISSUER_URL`, which you can obtain by following the documentation in [Obtain a JWT](../01-51-get-jwt.md). `DOMAIN_TO_EXPOSE_WORKLOADS` refers to either a custom domain or, as in this example, the default domain `local.kyma.dev`. Generate the `COOKIE_SECRET` using the following command:
> ```bash
> openssl rand -base64 32 | tr -- '+/' '-_'
> ```
> `TOKEN_SCOPES` specifies the OAuth scopes. Each provider has a default set of scopes that will be used if no custom scopes are configured.

Reviewer: ```suggestion
The example above shows the configuration of oauth2-proxy with the following parameters: 
- `CLIENT_SECRET`, `CLIENT_ID`, and `OIDC_ISSUER_URL`, which you can obtain by following the documentation in [Obtain a JWT](../01-51-get-jwt.md)
- `DOMAIN_TO_EXPOSE_WORKLOADS` refers to either a custom domain or, as in this example, the default domain `local.kyma.dev`
- `COOKIE_SECRET` that you can generate using the following command:
    ```bash
    openssl rand -base64 32 | tr -- '+/' '-_'
    ```
- `TOKEN_SCOPES` specifies the OAuth scopes. Each provider has a default set of scopes that are used if no custom scopes are configured.
```

For a complete list of options and further details, refer to the [oauth2-proxy documentation](https://oauth2-proxy.github.io/oauth2-proxy/configuration/overview/#config-options).

> To install `oauth2-proxy` with your configuration, use [oauth2-proxy helm chart](https://github.com/oauth2-proxy/manifests):

Reviewer: ```suggestion
   2. To install `oauth2-proxy` with your configuration, use [oauth2-proxy helm chart](https://github.com/oauth2-proxy/manifests):
```

  port: 8000
gateway: kyma-system/kyma-gateway
rules:
  - path: /anything
    methods:
      - POST
    accessStrategies:
      - handler: oauth2_introspection
        config:
          introspection_request_headers:
            Authorization: Basic {ENCODED_CREDENTIALS}
          introspection_url: https://{IAS_TENANT}.accounts.ondemand.com/oauth2/introspect
          required_scope:
            - write
  - path: /.*
    methods:
      - GET
    accessStrategies:
      - handler: oauth2_introspection
        config:
          introspection_request_headers:
            Authorization: Basic {ENCODED_CREDENTIALS}
          introspection_url: https://{IAS_TENANT}.accounts.ondemand.com/oauth2/introspect
          required_scope:
            - read
```
The above configuration uses the **oauth2_introspection** handler to expose HTTPBin service's `/anything` and `/.*` endpoints.

2. In order for the `extAuth` handler in APIRule `v2` to work, you must first deploy a service that will act as an external authorizer for Istio. This tutorial uses [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) with an OAuth2.0 compliant authorization server supporting OIDC discovery.

>    1. Replace the placeholders and create the `values.yaml` file with the oauth2-proxy configuration.

Reviewer 2: better to have 2.a 2.b. 2.c

Author: i adjusted these bullet points so now it renders correctly

Reviewer: I mean, we could try adding 2a 2b 2c, but there might be some problems with indenting the code blocks below (it won't be treated as a list, just regular text)
